roachtest: remove direct go calls

herkolategan opened the pull request:
Previously, tests had to use  bare goroutine calls to start a goroutine. This change removes all bare goroutine calls and replaces it with the new task APIs.

The use of `errgroup.Group` and `Monitor` still remains and will be addressed in a different PR.

Informs: #118214

Epic: None
Release note: None

herkolategan committed:
roachtest: remove direct go calls

Previously, tests had to use bare goroutine calls to start a goroutine. This
change removes all bare goroutine calls and replaces it with the new task APIs.
Most of the tests already handle errors from goroutines via a channel. This
logic is best left in its current state. New implementations can choose whether
they want to fail a test by returning an error from a managed goroutine, or
would rather handle errors some other way.

The use of `errgroup.Group` and `Monitor` still remains and will be addressed in
a different PR.

Informs: #118214

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/134644)
<!-- Reviewable:end -->

DarrylWong commented on pkg/cmd/roachtest/tests/cancel.go:
```diff
@@ -66,30 +68,32 @@ func registerCancel(r registry.Registry) {
 								close(sem)
-								return
+								// Errors are handled in the main goroutine.
+								return nil //nolint:returnerrcheck
```
nit: it's not immediately obvious that returning errors in `t.Go` fails the test. I think it makes sense for that to be the behavior but you have to go a couple layers and peek at the implementation to confirm that is the case.

srosenberg replied:
Agree with Darryl's comment; we should make this explicit in the interface doc. Also, this test could technically be refactored further to explicitly return `err` instead of passing it into `errCh`, but I also understand if you want to reduce the footprint of all changes in this PR.

herkolategan replied:
I had the same thought, will need to make it clear that returning an error from a task will result in a test failure. We could alternatively provide a version of the Go/GoWithCancel methods that does not take an error return.

From the footprint perspective, I didn't want to fiddle too much with already working code. But we could definitely make some of these more ergonomic. I'll create an issue for it, since engineers tend to borrow from existing code, it would be better if the examples are improved.

srosenberg replied:
> We could alternatively provide a version of the Go/GoWithCancel methods that does not take an error return.

No bother, it's not that hard to add `return nil` :)

srosenberg commented on pkg/cmd/roachtest/tests/allocator.go:
```diff
@@ -81,18 +82,10 @@ func registerAllocator(r registry.Registry) {
 		c.Run(ctx, option.WithNodes(c.Node(1)), "./cockroach workload init kv --drop {pgurl:1}")
 		for node := 1; node <= nodes; node++ {
-			node := node
-			// TODO(dan): Ideally, the test would fail if this queryload failed,
```
Ditto!

srosenberg commented on pkg/cmd/roachtest/tests/allocator.go:
```diff
@@ -81,18 +82,10 @@ func registerAllocator(r registry.Registry) {
-			go func() {
+			t.Go(func(taskCtx context.Context, _ *logger.Logger) error {
 				cmd := fmt.Sprintf("./cockroach workload run kv --tolerate-errors --min-block-bytes=8 --max-block-bytes=127 {pgurl%s}", c.Node(node))
-				l, err := t.L().ChildLogger(fmt.Sprintf(`kv-%d`, node))
```
Looks like this logger was previously unused?

herkolategan replied:
Yep, had to look twice before removing, and in actual fact was dead code.

srosenberg commented on pkg/cmd/roachtest/tests/cluster_init.go:
```diff
@@ -87,11 +88,11 @@ func runClusterInit(ctx context.Context, t test.Test, c cluster.Cluster) {
 		t.L().Printf("checking that the SQL conns are not failing immediately")
 		errCh := make(chan error, len(dbs))
 		for _, db := range dbs {
-			db := db
```
Good riddance :)

srosenberg reviewed (approved):
Thanks for refactoring all "naked" go calls! LGTM. We should probably do SELECT_PROBABILITY=0.6 to make sure we hit enough tests to shake out any potential regressions.

herkolategan commented on pkg/cmd/roachtest/tests/multitenant_upgrade.go:
```diff
@@ -280,16 +281,18 @@ func runMultitenantUpgrade(ctx context.Context, t test.Test, c cluster.Cluster)
+				return nil
+			})
 
 			wg.Wait()
```
Yes! I do see value in it, hence a follow-up PR, that adds group management to the task manager: https://github.com/cockroachdb/cockroach/pull/135831

herkolategan commented:
> Thanks for refactoring all "naked" go calls! LGTM. We should probably do SELECT_PROBABILITY=0.6 to make sure we hit enough tests to shake out any potential regressions.

Definitely, I ran a targeted job on all the affected job with a rather large regex. But I think a 0.6 probability would also make for a good final check. Will do one before merging today/tomorrow.

herkolategan committed:
roachtest: add Tasker error handling comment

Added a comment to inform test implementors, using the task API, that returning
an error from a Tasker goroutine will fail a test.

Informs: #118214

Epic: None
Release note: None

herkolategan commented:
TFTRs!

bors r=srosenberg,DarrylWong
